libgit/autogit_node_wrappers.go

```diff
@@ -33,16 +36,196 @@ import (
 //   a `tlfTypeNode`.
 
 type ctxReadWriteKeyType int
+type ctxSkipPopulateKeyType int
 
 const (
-	autogitRoot                         = ".kbfs_autogit"
-	ctxReadWriteKey ctxReadWriteKeyType = 1
+	autogitRoot = ".kbfs_autogit"
+
+	populateTimeout = 10 * time.Second
+
+	ctxReadWriteKey    ctxReadWriteKeyType    = 1
+	ctxSkipPopulateKey ctxSkipPopulateKeyType = 1
 
 	public  = "public"
 	private = "private"
 	team    = "team"
 )
 
+type repoNode struct {
+	libkbfs.Node
+	am       *AutogitManager
+	h        *libkbfs.TlfHandle
+	repoName string
+
+	lock                 sync.Mutex
+	populated            bool
+	populatingInProgress chan struct{}
+}
+
+var _ libkbfs.Node = (*repoNode)(nil)
+
+func newRepoNode(n libkbfs.Node, am *AutogitManager, h *libkbfs.TlfHandle,
+	repoName string) *repoNode {
+	return &repoNode{
+		Node:     n,
+		am:       am,
+		h:        h,
+		repoName: repoName,
+	}
+}
+
+func (rn *repoNode) dstDir() string {
+	var typeStr string
+	switch rn.h.Type() {
+	case tlf.Public:
+		typeStr = public
+	case tlf.Private:
+		typeStr = private
+	case tlf.SingleTeam:
+		typeStr = team
+	}
+
+	return path.Join(
+		autogitRoot, typeStr, string(rn.h.GetCanonicalName()))
+}
+
+func (rn *repoNode) populate(ctx context.Context) bool {
+	ctx = context.WithValue(ctx, ctxSkipPopulateKey, 1)
+	children, err := rn.am.config.KBFSOps().GetDirChildren(ctx, rn)
+	if err != nil {
+		rn.am.log.CDebugf(ctx, "Error getting children: %+v", err)
+		return false
+	}
+
+	h, err := rn.am.config.KBFSOps().GetTLFHandle(ctx, rn)
+	if err != nil {
+		rn.am.log.CDebugf(ctx, "Error getting handle: %+v", err)
+		return false
+	}
+
+	// If the directory is empty, clone it.  Otherwise, pull it.
+	var doneCh <-chan struct{}
+	cloneNeeded := len(children) == 0
```

Comment:
Yeah I think you're right on both points. I'd have to test it though.  It's possible the initial clone will actually fail if there no bare git repo, which would leave a CLONING file behind.  Might need to work out a better strategy at some point for communicating these background errors to the users, if this comes up in the real world.

Reply:
> which would leave a CLONING file behind

To make sure, this would cause next read to trigger a "pull" which is essentially a "reset", and from there it would start working again right?

Reply:
Not a read from the same instance, no.  Because `populate` gets set, a read won't trigger any more updates.  Though as of the next PR, a write to the repo would trigger a reset and override the CLONING file, if the reset is successful.  So basically, if the clone fails, it will look stuck until the next write or until the next KBFS instance/restart tries to read.  This is kind of crappy behavior, but I'd argue we should fix it separately after this gets merged, if we want to and/or if it becomes a real issue.

Reply:
Sounds good! I don't think an empty repo will be a very common case for autogit anyway.